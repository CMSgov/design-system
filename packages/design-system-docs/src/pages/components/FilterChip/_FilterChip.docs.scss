/*
Filter Chip

Filter Chips are used to display dismissible filter chips

@status Draft

Markup: filterchip.example.html

Style guide: components.filterchip
*/

/*
`<FilterChip>`

@react-example FilterChip.example.jsx

@react-props FilterChip.tsx

Style guide: components.filterchip.react
*/

/*
---

### When to use

- Use to clearly indicate to the user what filters have been applied to a set of results and provide a quick and easy way to remove a selected filter

### When to consider alternatives

- When many filter chips would be displayed and available screen real estate is limited
- When a badge is sufficient in place of a filter chip

### Usage

- Do not mix badges and filter chip as it can be confusing to the user
- Do not use filter chip when a badge is sufficient, ie when none of the tags are removable
- Filter chips ideally should be short and not wrap to the second line
- Allow adequate spacing (at least 8px) between each filter chip, as much as the design will allow.
- The colors of the Filter Chip can be overriden using the Sass variables listed below, but should not be overriden with colors outside of the color palette.

### Accessibility

- Use `ariaClearLabel` prop so that screenreaders can accurately read out the close icon action
- `Enter`, `Backspace`, `Delete`, and `Spacebar` will remove the tag when it is in focus

### Future research

### Customization

The following Sass variables can be overridden to the filter chip:
- `$filter-chip-icon-container-size` - Height / Width for the close icon container
- `$filter-chip-icon-size` - Height / Width for the close icon in the filter chip
- `$filter-chip-border-radius` - Border radius for the filter chip
- `$filter-chip-text-color` - Text color for the filter chip
- `$filter-chip-background-color` - Background color for the filter chip
- `$filter-chip-border-color` - Border color for the filter chip
- `$filter-chip-text-color-hover` - Text color for the filter chip on hover
- `$filter-chip-background-color-hover` -  Background color for the filter chip on hover
- `$filter-chip-border-color-hover` -  Border color for the filter chip on hover

Style guide: components.filterchip.guidance
*/
