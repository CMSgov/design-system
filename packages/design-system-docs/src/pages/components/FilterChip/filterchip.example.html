<button class="ds-c-filter-chip__button" id="filter_1">
  <span class="ds-c-filter-chip__label">Example Filter Chip (HTML Only)</span>
  <span class="ds-c-filter-chip__clear-icon-container" aria-label="Remove">
    <svg
      aria-labelledby="icon-2__title"
      aria-hidden="false"
      class="ds-c-icon ds-c-icon--close"
      focusable="false"
      id="icon-2"
      role="img"
      viewBox="0 0 16 16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title id="icon-2__title">Close</title>
      <path
        d="M14.647 11.213c.235.235.353.521.353.858 0 .337-.118.624-.353.859l-1.717 1.717a1.17 1.17 0 01-.86.354c-.336 0-.622-.118-.857-.354l-3.714-3.712-3.712 3.712A1.166 1.166 0 012.93 15c-.337 0-.622-.118-.859-.354L.353 12.93A1.165 1.165 0 010 12.07c0-.337.117-.623.353-.858L4.065 7.5.353 3.789A1.168 1.168 0 010 2.929c0-.336.117-.622.353-.857L2.07.353C2.307.118 2.592 0 2.93 0c.337 0 .623.118.858.353L7.5 4.065 11.213.353c.235-.235.521-.353.857-.353.337 0 .623.118.86.353l1.717 1.719c.235.235.353.521.353.857 0 .338-.118.623-.353.86L10.935 7.5l3.712 3.712z"
        fill-rule="evenodd"
      ></path>
    </svg>
  </span>
</button>
