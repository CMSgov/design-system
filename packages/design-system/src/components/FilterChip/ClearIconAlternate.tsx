import React from 'react';

/**
 * <ClearIconAlternate> is an internal component used by <FilterChip>
 */

export function ClearIconAlternate(): React.ReactElement {
  return (
    <svg
      className="ds-c-clear-icon"
      height="14"
      viewBox="0 0 14 14"
      width="14"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path d="m8.6002309 7 5.2573243-5.25732429c.1899264-.1899264.1899264-.49704142 0-.68696782l-.9132631-.91326309c-.1899264-.1899264-.4970414-.1899264-.6869678 0l-5.2573243 5.2573243-5.25732429-5.2573243c-.1899264-.1899264-.49704142-.1899264-.68696782 0l-.91326309.91326309c-.1899264.1899264-.1899264.49704142 0 .68696782l5.2573243 5.25732429-5.2573243 5.2573243c-.1899264.1899264-.1899264.4970414 0 .6869678l.91326309.9132631c.1899264.1899264.49704142.1899264.68696782 0l5.25732429-5.2573243 5.2573243 5.2573243c.1899264.1899264.4970414.1899264.6869678 0l.9132631-.9132631c.1899264-.1899264.1899264-.4970414 0-.6869678z" />
    </svg>
  );
}

export default ClearIconAlternate;
