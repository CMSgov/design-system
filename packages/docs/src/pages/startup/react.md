---
title: Using React components
weight: 3
---

The examples below assume you've installed the design system using NPM and have already setup your build system.

<h4 class="ds-h4 ds-u-font-size--base">Default imports</h4>

Individual components can be imported from their individual export file.

```jsx
import Button from '@cmsgov/design-system-core/dist/components/Button/Button';
```

<h4 class="ds-h4 ds-u-font-size--base">Named imports</h4>

Components can also be imported using the shorter syntax below.

<div class="ds-c-alert ds-c-alert--warn ds-u-margin-bottom--2 ds-u-font-size--small">
  <div class="ds-c-alert__body">
    <h3 class="ds-c-alert__heading ds-u-font-size--base">Performance note</h3>
    <p class="ds-c-alert__text">
      This approach may result in a much larger file than you intend. Depending on what module bundler you use, all of the design system's React components may be included in the bundled file even if you didn't specifically import them. This can be avoided by enabling features like [tree shaking in Webpack](https://webpack.js.org/guides/tree-shaking/).
    </p>
  </div>
</div>

```jsx
import { Button } from '@cmsgov/design-system-core';
```

[View an example](https://github.com/CMSgov/design-system/blob/master/examples/react-app/src/scripts/index.js)

<h2 id="examples" class="ds-h2 ds-u-color--primary-darker">Examples</h2>

Additional examples of the design system in use can be viewed on GitHub. These projects demonstrate the various ways you can incorporate the design system into your development process and various use cases.

<a href="https://github.com/CMSgov/design-system/tree/master/examples/" class="ds-c-button">Browse example projects</a>

<h2 id="need-help" class="ds-h2 ds-u-color--primary-darker">Need help or ran into an issue?</h2>

If you're having trouble installing or setting up the design system, or if you think you've found a bug, feel free to [open an issue on GitHub](https://github.com/CMSgov/design-system/issues).
