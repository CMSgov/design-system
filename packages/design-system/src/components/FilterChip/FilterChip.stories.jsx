import FilterChip from './FilterChip';
import { action } from '@storybook/addon-actions';

export default {
  title: 'Components/Filter Chip',
  component: FilterChip,
  argTypes: {
    size: {
      control: { type: 'radio' },
      options: ['big', 'default'],
    },
  },
  args: {
    label: 'Example Filter Chip',
    size: 'default',
  },
};

const Template = ({ data, ...args }) => <FilterChip {...args} />;

export const SingleChip = Template.bind({});

export const MultipleChips = () => {
  const listOfChips = [
    { label: 'Example FilterChip' },
    { label: 'Example with alternate icon', useAlternateIcon: true },
    { label: 'Example big filter chip', size: 'big' },
    { label: 'Example big with alternate icon', useAlternateIcon: true, size: 'big' },
  ];

  const onDelete = action('onDelete');

  return listOfChips.map((chipData) => (
    <FilterChip key={chipData.label} {...chipData} onDelete={onDelete} />
  ));
};
