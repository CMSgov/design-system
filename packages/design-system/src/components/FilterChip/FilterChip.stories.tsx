import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import { FilterChip } from './FilterChip';
import { action } from '@storybook/addon-actions';

const meta: Meta<typeof FilterChip> = {
  title: 'Components/FilterChip',
  component: FilterChip,
  args: {
    label: 'Example Filter Chip',
  },
};
export default meta;

type Story = StoryObj<typeof FilterChip>;

export const Default: Story = {};

export const MultipleChips: Story = {
  render: function Component() {
    const listOfChips = [
      { label: 'Example FilterChip' },
      { label: 'Example with alternate icon', useAlternateIcon: true },
      { label: 'Example big filter chip', size: 'big' },
      { label: 'Example big with alternate icon', useAlternateIcon: true, size: 'big' },
    ];

    const onDelete = action('onDelete');

    return (
      <>
        {listOfChips.map((chipData) => (
          <FilterChip key={chipData.label} {...(chipData as any)} onDelete={onDelete} />
        ))}
      </>
    );
  },
};
