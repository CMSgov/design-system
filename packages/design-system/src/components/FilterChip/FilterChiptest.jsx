import FilterChip from './FilterChip';
import React from 'react';
import { shallow } from 'enzyme';

describe('FilterChip', () => {
  it('should include children as label', () => {
    expect(shallow(<FilterChip label="Foo" />).text()).toEqual(' Foo<ClearIcon />');
  });

  it('renders button and should disappear when clicked', () => {
    const wrapper = shallow(<FilterChip label="Foo" />);
    expect(wrapper.text()).toEqual(' Foo<ClearIcon />');

    const button = wrapper.find('button');
    expect(button.length).toEqual(1);

    button.simulate('click');
    wrapper.update();
    expect(wrapper.text()).toEqual('');
  });
});
