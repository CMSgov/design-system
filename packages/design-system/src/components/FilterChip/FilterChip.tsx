import { ClearIcon } from '../ClearIcon';
import { ClearIconAlternate } from './ClearIconAlternate';
import React from 'react';
import classNames from 'classnames';
import uniqueId from 'lodash.uniqueid';

const actionableKeys = ['Enter', 'Backspace', 'Delete'];

export interface FilterChipProps {
  /**
   * Id for filter chip button. If not provided, a unique id will be automatically generated and used.
   */
  id?: string;
  /**
   * Class to be applied to the outer `<div>` that contains filter chip button.
   */
  className?: string;
  /**
   * Text for the filter chip
   */
  label: string;
  /**
   *  For screenreaders, text to read for removal
   */
  ariaClearLabel: string;
  /**
   * Function to call when filter chip is dismissed
   */
  onDelete: () => void;
  /**
   *  Use alternate thinner close icon in place of standard
   */
  useAlternateIcon?: boolean;
  /**
   * Sets the size of the chip to larger version
   */
  size?: 'big';
}

export class FilterChip extends React.Component<FilterChipProps> {
  constructor(props: FilterChipProps) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
  }

  handleClick(): void {
    this.props.onDelete();
  }

  handleKeyDown(evt: React.KeyboardEvent): void {
    if (actionableKeys.includes(evt.key)) {
      this.handleClick();
      evt.preventDefault();
    }
  }

  render(): React.ReactNode {
    const { id, label, ariaClearLabel, className, useAlternateIcon, size } = this.props;
    const buttonClassNames = classNames(
      'ds-c-filter-chip__button',
      size && size === 'big' ? 'ds-c-filter-chip__button--big' : '',
      className
    );
    const iconContainerClassNames = classNames(
      'ds-c-filter-chip__clear-icon-container',
      useAlternateIcon ? 'ds-c-filter-chip__clear-icon-alternate-container' : ''
    );
    return (
      <>
        <button
          className={buttonClassNames}
          id={id || uniqueId(`filter_`)}
          onClick={this.handleClick}
          onKeyDown={this.handleKeyDown}
        >
          <span className="ds-c-filter-chip__label">{label}</span>
          <span className={iconContainerClassNames} aria-label={ariaClearLabel}>
            {useAlternateIcon ? <ClearIconAlternate /> : <ClearIcon />}
          </span>
        </button>
      </>
    );
  }
}

export default FilterChip;
