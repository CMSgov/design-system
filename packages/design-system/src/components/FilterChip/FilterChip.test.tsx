import React from 'react';
import FilterChip from './FilterChip';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('FilterChip', () => {
  const renderFilterChip = (customProps = {}) => {
    const defaultProps = {
      label: 'Foo',
      onDelete: jest.fn(),
      id: 'static-id',
    };
    const props = { ...defaultProps, ...customProps };
    return render(<FilterChip {...props} />);
  };

  it('should include children as label', () => {
    renderFilterChip();
    const chipEl = screen.getByRole('button');
    expect(chipEl.querySelector('.ds-c-filter-chip__label').textContent).toEqual('Foo');
    expect(chipEl).toMatchSnapshot();
  });

  it('should use different aria label if provided', () => {
    renderFilterChip({ ariaClearLabel: 'Clear' });
    const chipEl = screen.getByRole('button');
    expect(chipEl.querySelector('.ds-u-visibility--screen-reader').textContent).toEqual(
      'Clear Foo filter .'
    );
  });

  it('generates an id when no id is provided', () => {
    renderFilterChip({ id: undefined });
    const idRegex = /filter-chip--\d+/;
    expect(screen.getByRole('button').id).toMatch(idRegex);
  });

  describe('onDelete', () => {
    it('should call onDelete function when clicked', () => {
      const onDelete = jest.fn();
      renderFilterChip({ onDelete: onDelete });

      const chipEl = screen.getByRole('button');
      expect(chipEl).toBeDefined();

      userEvent.click(chipEl);
      expect(onDelete).toHaveBeenCalled();
    });

    it('should call onDelete when certain keyboard keys are pressed', () => {
      const onDelete = jest.fn();
      renderFilterChip({ onDelete: onDelete });

      const chipEl = screen.getByRole('button');
      chipEl.focus();

      userEvent.keyboard('{Enter}');
      expect(onDelete).toHaveBeenCalledTimes(1);
      userEvent.keyboard('{Space}');
      expect(onDelete).toHaveBeenCalledTimes(2);
      userEvent.keyboard('{Backspace}');
      expect(onDelete).toHaveBeenCalledTimes(3);
      userEvent.keyboard('{Delete}');
      expect(onDelete).toHaveBeenCalledTimes(4);
    });

    it('should not call onDelete when most keyboard keys are pressed', () => {
      const onDelete = jest.fn();
      renderFilterChip({ onDelete: onDelete });

      const chipEl = screen.getByRole('button');
      chipEl.focus();
      userEvent.keyboard('{Tab}');
      expect(onDelete).not.toHaveBeenCalled();
      userEvent.keyboard('a');
      expect(onDelete).not.toHaveBeenCalled();
    });
  });

  it('should set big CSS class', () => {
    renderFilterChip({ size: 'big' });
    const chipEl = screen.getByRole('button');
    expect(chipEl.classList).toContain('ds-c-filter-chip__button--big');
  });

  it('should use alternate icon', () => {
    renderFilterChip({ useAlternateIcon: true });
    const chipEl = screen.getByRole('button');
    const iconContainerEl = chipEl.querySelector('.ds-c-filter-chip__clear-icon-container');

    expect(iconContainerEl).toBeDefined();
    expect(iconContainerEl.classList).toContain('ds-c-filter-chip__clear-icon-alternate-container');
    expect(iconContainerEl).toMatchSnapshot();
  });
});
