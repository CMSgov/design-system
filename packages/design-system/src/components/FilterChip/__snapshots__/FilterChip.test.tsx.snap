// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterChip should include children as label 1`] = `
<button
  class="ds-c-filter-chip__button"
  id="static-id"
  type="button"
>
  <span
    aria-describedby="static-id-instructions"
    aria-hidden="true"
    class="ds-c-filter-chip__label"
  >
    Foo
  </span>
  <span
    class="ds-u-visibility--screen-reader"
    id="static-id-instructions"
  >
    Remove
     
    Foo filter
     .
  </span>
  <span
    class="ds-c-filter-chip__clear-icon-container"
  >
    <svg
      aria-hidden="true"
      class="ds-c-icon ds-c-icon--close "
      viewBox="0 0 16 16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M14.647 11.213c.235.235.353.521.353.858 0 .337-.118.624-.353.859l-1.717 1.717a1.17 1.17 0 01-.86.354c-.336 0-.622-.118-.857-.354l-3.714-3.712-3.712 3.712A1.166 1.166 0 012.93 15c-.337 0-.622-.118-.859-.354L.353 12.93A1.165 1.165 0 010 12.07c0-.337.117-.623.353-.858L4.065 7.5.353 3.789A1.168 1.168 0 010 2.929c0-.336.117-.622.353-.857L2.07.353C2.307.118 2.592 0 2.93 0c.337 0 .623.118.858.353L7.5 4.065 11.213.353c.235-.235.521-.353.857-.353.337 0 .623.118.86.353l1.717 1.719c.235.235.353.521.353.857 0 .338-.118.623-.353.86L10.935 7.5l3.712 3.712z"
        fill-rule="evenodd"
      />
    </svg>
  </span>
</button>
`;

exports[`FilterChip should use alternate icon 1`] = `
<span
  class="ds-c-filter-chip__clear-icon-container ds-c-filter-chip__clear-icon-alternate-container"
>
  <svg
    aria-hidden="true"
    class="ds-c-icon ds-c-icon--close ds-c-icon--close-thin "
    viewBox="-2 -2 18 18"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 13.0332964L13.0332964 0M13.0332964 13.0332964L0 0"
      fill="none"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-width="2"
    />
  </svg>
</span>
`;
